hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotRegionProcedure.java

```diff
@@ -0,0 +1,268 @@
+      LOG.info("This procedure {} is already finished, skip the rest processes", this.getProcId());
+      return;
+    }
+    if (event == null) {
+      LOG.warn("procedure event for {} is null, maybe the procedure is created when recovery",
+        getProcId());
+      return;
+    }
+    if (error == null) {
+      LOG.info("finish snapshot {} on region {}", snapshot.getName(), region.getEncodedName());
+      succ = true;
+    }
+
+    event.wake(env.getProcedureScheduler());
+    event = null;
+  }
+
+  @Override
+  public TableName getTableName() {
+    return region.getTable();
+  }
+
+  @Override
+  public TableOperationType getTableOperationType() {
+    return TableOperationType.REGION_SNAPSHOT;
+  }
+
+  @Override
+  protected Procedure<MasterProcedureEnv>[] execute(MasterProcedureEnv env)
+    throws ProcedureYieldException, ProcedureSuspendedException, InterruptedException {
+    if (dispatched) {
+      if (succ) {
+        return null;
+      }
+      dispatched = false;
+    }
+
+    RegionStates regionStates = env.getAssignmentManager().getRegionStates();
+    RegionStateNode regionNode = regionStates.getRegionStateNode(region);
+    regionNode.lock();
+    try {
+      if (regionNode.getProcedure() != null) {
+        setTimeoutForSuspend(env, String.format("region %s has a TRSP attached %s",
+          region.getRegionNameAsString(), regionNode.getProcedure()));
+        throw new ProcedureSuspendedException();
+      }
+      if (!regionNode.getState().matches(RegionState.State.OPEN)) {
+        setTimeoutForSuspend(env, String.format("region state of %s is %s",
+          region.getRegionNameAsString(), regionNode.getState()));
+        throw new ProcedureSuspendedException();
+      }
+      ServerName targetServer = regionNode.getRegionLocation();
+      if (targetServer == null) {
+        setTimeoutForSuspend(env, String.format("target server of region %s is null",
+          region.getRegionNameAsString()));
+        throw new ProcedureSuspendedException();
+      }
+      ServerState serverState = regionStates.getServerNode(targetServer).getState();
+      if (serverState != ServerState.ONLINE) {
+        setTimeoutForSuspend(env, String.format("target server of region %s %s is in state %s",
+          region.getRegionNameAsString(), targetServer, serverState));
+        throw new ProcedureSuspendedException();
+      }
+      try {
+        env.getRemoteDispatcher().addOperationToNode(targetServer, this);
+        dispatched = true;
+        event = new ProcedureEvent<>(this);
+        event.suspendIfNotReady(this);
+        throw new ProcedureSuspendedException();
+      } catch (FailedRemoteDispatchException e) {
+        setTimeoutForSuspend(env, "Failed send request to " + targetServer);
+        throw new ProcedureSuspendedException();
+      }
+    } finally {
+      regionNode.unlock();
+    }
+  }
+
+  @Override
+  protected void rollback(MasterProcedureEnv env) {
+    // nothing to rollback
```

Comment:
Better throw UnsupportedOperationException here as it seems that this procedure will never be rolled back?

Reply:
Ok, I will address it. Thanks, Duo.